usr/lib/api/statistics.c

```diff
@@ -0,0 +1,265 @@
+
+    mech_idx = mechtable_idx_from_numeric(mech->mechanism);
+    if (mech_idx < 0)
+        return CKR_MECHANISM_INVALID;
+
+    ofs += mech_idx * (NUM_SUPPORTED_STRENGTHS + 1) * sizeof(counter_t);
+
+    strength_idx = NUM_SUPPORTED_STRENGTHS - strength_idx;
+    ofs += strength_idx * sizeof(counter_t);
+
+    if (ofs > statistics->shm_size)
+        return CKR_FUNCTION_FAILED;
+
+    counter = (counter_t*)(statistics->shm_data + ofs);
+    __sync_add_and_fetch(counter, 1);
+
+    return CKR_OK;
+}
+
+/*
+ * Open the statistics shared memory segment for the specified user.
+ * If user is -1, then it is opened for the current user.
+ * If create is TRUE then the shared memory segment is created if it is not
+ * already existent.
+ */
+static CK_RV statistics_open_shm(struct statistics *statistics, int user,
+                                 CK_BBOOL create)
+{
+    int i, err, clear = 0;
+    struct stat stat_buf;
+
+    snprintf(statistics->shm_name, sizeof(statistics->shm_name) - 1,
+             "%s_stats_%u", CONFIG_PATH, user == -1 ? geteuid() : (uid_t)user);
+    for (i = 1; statistics->shm_name[i] != '\0'; i++) {
+        if (statistics->shm_name[i] == '/')
+            statistics->shm_name[i] = '.';
+    }
+    if (statistics->shm_name[0] != '/') {
+        memmove(&statistics->shm_name[1], &statistics->shm_name[0],
+                strlen(statistics->shm_name) + 1);
+        statistics->shm_name[0] = '/';
+    }
+
+    TRACE_INFO("Statistics SHM name: '%s'\n", statistics->shm_name);
+
+    statistics->shm_handle = shm_open(statistics->shm_name, O_RDWR,
+                                      S_IRUSR | S_IWUSR);
+    if (statistics->shm_handle == -1) {
+        if (create) {
+            /* try to create it */
+            statistics->shm_handle = shm_open(statistics->shm_name,
+                                              O_CREAT | O_RDWR,
+                                              S_IRUSR | S_IWUSR);
+            if (statistics->shm_handle == -1) {
+                err = errno;
+                TRACE_ERROR("Failed to create SHM '%s': %s\n",
+                            statistics->shm_name,  strerror(err));
+                OCK_SYSLOG(LOG_ERR, "Failed to create SHM '%s': %s\n",
+                           statistics->shm_name, strerror(err));
+                return CKR_FUNCTION_FAILED;
+            }
+
+            if (fchmod(statistics->shm_handle, S_IRUSR | S_IWUSR) == -1) {
+                err = -errno;
+                TRACE_ERROR("Failed to change mode of SHM '%s': %s\n",
+                            statistics->shm_name,  strerror(err));
+                OCK_SYSLOG(LOG_ERR, "Failed to change mode of SHM '%s': %s\n",
+                           statistics->shm_name, strerror(err));
+                close(statistics->shm_handle);
+                shm_unlink(statistics->shm_name);
+                return CKR_FUNCTION_FAILED;
+            }
+            if (fchown(statistics->shm_handle, user, -1) != 0) {
```

Comment:
Why is this needed?  It can only succeed if either `user` is already the owner of the SHM segment, or the current application runs as root.

Reply:
yes, you are right, will remove the fchown block.

Reply:
Maybe add this back in? This function is actually never called for another user, but if it would, then we should do the fchown for that user I guess.

Reply:
@juergenchrist Do you want to comment on this? 
Other than that I would say that this PR is ready to merge.

Reply:
Currently I see this function as `static` with exactly one caller who passes `-1` as user.  In general: what should be accomplished with the `fchown`?  Just sounds weird to me that one user might create the SHM-User-Segment of a different user.  The combination "read access" and "different user" makes sense, but then the segment should not be created.

Reply:
The original idea was that this function could be called by pkcsstats too open the segment of any user (when running as root), but it turned of that linking this module with the tool would cause too much trouble (many other unresolved externals), so it is only used at once place at the moment.  So yes, lets keep it as is.